File: lib/models/user_model.dart
class User {
  final String id;
  final String name;
  final String email;
  final String? avatar;
  final String? department;
  final String? studentId;
  final String? phone;
  final String role;
  final List<String> favoriteBooks;
  final DateTime createdAt;

  User({
    required this.id,
    required this.name,
    required this.email,
    this.avatar,
    this.department,
    this.studentId,
    this.phone,
    required this.role,
    required this.favoriteBooks,
    required this.createdAt,
  });

  factory User.fromJson(Map<String, dynamic> json) {
    List<String> favorites = [];
    if (json['favoriteBooks'] != null) {
      if (json['favoriteBooks'] is List) {
        favorites = (json['favoriteBooks'] as List)
            .map((item) => item is String ? item : item['_id'].toString())
            .toList()
            .cast<String>();
      }
    }

    return User(
      id: json['_id'],
      name: json['name'],
      email: json['email'],
      avatar: json['avatar'],
      department: json['department'],
      studentId: json['studentId'],
      phone: json['phone'],
      role: json['role'] ?? 'user',
      favoriteBooks: favorites,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'email': email,
      'avatar': avatar,
      'department': department,
      'studentId': studentId,
      'phone': phone,
      'role': role,
      'favoriteBooks': favoriteBooks,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}

File: backend/src/models/user.model.js
const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');
const validator = require('validator');

const userSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, 'Please provide your name'],
    trim: true,
    minlength: [3, 'Name must be at least 3 characters long'],
    maxlength: [50, 'Name cannot exceed 50 characters']
  },
  email: {
    type: String,
    required: [true, 'Please provide your email'],
    unique: true,
    lowercase: true,
    trim: true,
    validate: [validator.isEmail, 'Please provide a valid email']
  },
  password: {
    type: String,
    required: [true, 'Please provide a password'],
    minlength: [8, 'Password must be at least 8 characters long'],
    select: false // Don't return password in queries
  },
  avatar: {
    type: String,
    default: 'default-avatar.png'
  },
  department: {
    type: String,
    trim: true
  },
  studentId: {
    type: String,
    trim: true
  },
  phone: {
    type: String,
    trim: true
  },
  role: {
    type: String,
    enum: ['user', 'admin'],
    default: 'user'
  },
  favoriteBooks: [{
    type: mongoose.Schema.ObjectId,
    ref: 'Book'
  }],
  createdAt: {
    type: Date,
    default: Date.now
  }
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Virtual field for user's listings
userSchema.virtual('listings', {
  ref: 'Book',
  foreignField: 'owner',
  localField: '_id'
});

// Hash password before saving
userSchema.pre('save', async function(next) {
  // Only hash the password if it's modified or new
  if (!this.isModified('password')) return next();
  
  try {
    const salt = await bcrypt.genSalt(10);
    this.password = await bcrypt.hash(this.password, salt);
    next();
  } catch (error) {
    next(error);
  }
});

// Method to compare password
userSchema.methods.comparePassword = async function(candidatePassword) {
  return await bcrypt.compare(candidatePassword, this.password);
};

module.exports = mongoose.model('User', userSchema); 